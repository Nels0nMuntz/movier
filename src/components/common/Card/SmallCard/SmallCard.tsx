import React from "react";
import Stack from "@mui/material/Stack/Stack";
import { Link } from "react-router-dom";
import { faInfo, faPlus } from "@fortawesome/free-solid-svg-icons";
import { observer } from "mobx-react-lite";

import { Action, Poster, StyledLink, Wrapper } from "./styled";
import { getW300ImageUrl } from "api";
import { Typography } from "../../Typography/Typography";
import { SkeletonProvider, Skeleton } from "../../Skeleton";
import { FAIcon } from "../../FAIcon/FAIcon";

import noImage from "../../../../assets/img/no-image.svg"


interface Props {
  title: string;
  poster_path: string | null;
  release_date: string;
  adult: boolean;
  genres: string[];
  sourcePath: string;
}

export const SmallCard: React.FC<Props> = observer((props) => {

  const { title, adult, poster_path, release_date, genres, sourcePath } = props;

  const [ready, setReady] = React.useState(true);

  let Img;

  if (poster_path) {
    const src = getW300ImageUrl(poster_path);
    Img = <img src={src} alt={title} width={272} height={408} />
  } else {
    Img = <img src={noImage} alt={title} width={272} height={408} />
  }

  const genresString = genres.length ? genres.slice(0, 3).join(" / ") : "";

  React.useEffect(() => {
    if (poster_path) {
      const src = getW300ImageUrl(poster_path);
      const buffer = new Image();
      buffer.onload = () => setReady(true);
      buffer.src = src;
    }
  }, [])

  return (
    <SkeletonProvider visible={!ready}>
      <Wrapper>
        <Stack direction="column" gap={2}>
          <Poster>
            <Skeleton variant="rectangular">
              {Img}
            </Skeleton>
            {ready && (
              <Stack direction="column" className="actions" gap={1} component="ul">
                <li>
                  <Link to="/" aria-label="Add to my list" title="Add to my list">
                    <Action>
                      <FAIcon icon={faPlus} />
                    </Action>
                  </Link>
                </li>
                <li>
                  <Link to={sourcePath} aria-label="Read more" title="Read more">
                    <Action>
                      <FAIcon icon={faInfo} />
                    </Action>
                  </Link>
                </li>
              </Stack>
            )}
          </Poster>
          <Skeleton variant="text">
            <StyledLink to={sourcePath}>
              <Typography element="span" type="heading_6">{title}</Typography>
            </StyledLink>
          </Skeleton>
          <Skeleton variant="text">
            <Typography element="span" type="body_1" color="secondary">{genresString}</Typography>
          </Skeleton>
          <Skeleton variant="text">
            <Stack direction="row" gap={2}>
              <Typography element="span" type="body_1" color="secondary">{release_date.slice(0, 4)}</Typography>
              {adult && <Typography element="span" type="body_1" color="secondary">18+</Typography>}
            </Stack>
          </Skeleton>
        </Stack>
      </Wrapper>
    </SkeletonProvider>
  )
});
