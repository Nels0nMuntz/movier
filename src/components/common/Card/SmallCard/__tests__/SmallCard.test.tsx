import { Props, SmallCard } from "../SmallCard";
import { BrowserRouter } from "react-router-dom";
import { shallow } from "enzyme";
import { log } from "console";


let props: Partial<Props> = {};

const renderComponent = () => shallow(
  <BrowserRouter>
    <SmallCard {...props as Props} />
  </BrowserRouter>
)

describe("SmallCard", () => {
  beforeEach(() => {
    props = {
      adult: false,
      genres: ["Action", "Trailer"],
      poster_path: "https://image.tmdb.org/t/p/w300/zsbolOkw8RhTU4DKOrpf4M7KCmi.jpg",
      release_date: "2023-07-06",
      sourcePath: "/movie/457332",
      title: "Hidden Strike"
    }
  });
  it("renders correctly", () => {
    expect(renderComponent()).toMatchSnapshot()
  });
  describe("when `adult` prop is passed", () => {
    beforeEach(() => {
      props.adult = true
    });
    it("renders `18+` text", () => {
      expect(renderComponent().find({ children: "18+" })).toHaveLength(1)
    });
  });
  // describe("when `adult` prop is not passed", () => {
  //   it("does not render `18+` text", () => {
  //     expect(renderComponent().find({ children: "18+" }).html()).not.toBeInTheDocument()
  //   });
  // });
})