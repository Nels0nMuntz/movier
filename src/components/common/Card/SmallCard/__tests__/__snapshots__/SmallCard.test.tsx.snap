// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallCard renders correctly 1`] = `ShallowWrapper {}`;
